### PR TITLE
Pragma linter

### PR DESCRIPTION
**Pull request by n-wbrown**
I've added the more aggressive pragma linting features in this commit. I think we both added the xgmd project but in different ways so after my wild git-rebasing adventure to make this mergeable it looks like we broke a few other tests.

### CODE REVIEW

File: pytmc/bin/pragmalint.py [hunk 1]
```diff
@@ -21,9 +21,12 @@
 
 PRAGMA_START_RE = re.compile('{attribute')
 PRAGMA_RE = re.compile(
-    r"^{\s*attribute[ \t]+'pytmc'[ \t]*:=[ \t]*'([\s\S]*)'}$",
+    r"^{\s*attribute[ \t]+'pytmc'[ \t]*:=[ \t]*'(?P<setting>[\s\S]*)'}$",
     re.MULTILINE
 )
+PRAGMA_LINE_RE = re.compile(r"([^;'\n]+)", re.MULTILINE)
+PRAGMA_SETTING_RE = re.compile(r"^\s*(?P<title>[a-zA-z0-9]+):(?P<setting>.*)$")
```
Comment (klauer) [inline pytmc/bin/pragmalint.py]: Maybe we could be a bit more lax on whitespace near the "title", and strip the whitespace before/after "setting"?
Something like `r"^\s*(?P<title>[a-zA-z0-9]+)\s*:\s*(?P<setting>\S*)\s*$`

File: pytmc/bin/pragmalint.py [hunk 2]
```diff
@@ -105,9 +137,20 @@ def _build_map_of_offset_to_line_number(source):
...
+    ----------
+    filename : str
+        Target file name to be linted.
+
+    source : subclass of pytmc.parser.TwincatItem
+        Representation of TwinCAT project chunk in which to search for the
+        filename argument
+
+    verbose : bool
+        ???
```
Comment (klauer) [inline pytmc/bin/pragmalint.py]:
```suggestion
        Show more context around the linting process, including all source file names and
        number of pragmas found
```
:)

File: pytmc/bin/pragmalint.py [hunk 3]
```diff
@@ -84,6 +87,35 @@ def lint_pragma(pragma):
...
+        line_match = PRAGMA_SETTING_RE.match(line)
+        if line_match:
+            config_lines_detected += 1
+            pv_match = PRAGMA_PV_LINE_RE.search(line)
+            if pv_match:
+                pv_line_detected += 1
+
+    # There shall be at one config line at minimum
+    # There shall be a config line for pv even if it's just "pv:"
+    if (config_lines_detected <= 0 or pv_line_detected != 1):
```
Comment (klauer) [inline pytmc/bin/pragmalint.py]: We can have more than one pv line though, right?

File: pytmc/bin/pragmalint.py (hunk 1 above)
Comment (klauer) [reply pytmc/bin/pragmalint.py]: Also, your original regexes are here: https://github.com/slaclab/pytmc/blob/6a35be09f5b5e9ab47d14845732b32a370c8e827/pytmc/pragmas.py#L16

Wonder if we should refer to at least one of them here (or fix those up?)

File: pytmc/bin/pragmalint.py (hunk 2 above)
Comment (n-wbrown) [reply pytmc/bin/pragmalint.py]: Thanks @klauer

File: pytmc/bin/pragmalint.py (hunk 3 above)
Comment (n-wbrown) [reply pytmc/bin/pragmalint.py]: Good point. Fixing now

File: pytmc/bin/pragmalint.py (hunk 1 above)
Comment (n-wbrown) [reply pytmc/bin/pragmalint.py]: Good point. Using the common specification and updating it.  

With regards to the whitespace handling, I think `strip` is a more thorough solution. Some of our settings actually feature spaces inside so the `\S` should break e.g. on `field: ZNAM OFF`. If we still use the `.` inside greedy implementation means that the last `\s` won't be sought and we'll still need to `strip` the string later.

File: pytmc/bin/pragmalint.py [hunk 4]
```diff
@@ -84,6 +87,35 @@ def lint_pragma(pragma):
...
+        pragma_setting = match.groupdict()["setting"]
+    except KeyError:
+        # if no configuration region in the pragma was detected then fail
+        raise LinterError()
+
+    config_lines = PRAGMA_LINE_RE.findall(pragma_setting)
+    if len(config_lines) == 0:
+        # It is not acceptable to lack configuration lines
+        # At a minimum "pv: " must exist
+        raise LinterError()
```
Comment (klauer) [inline pytmc/bin/pragmalint.py]: Let's move these comments into the `LinterError` message itself - it'll help the user determine what's wrong with the pragma.

File: pytmc/bin/pragmalint.py [hunk 5]
```diff
@@ -84,6 +87,35 @@ def lint_pragma(pragma):
...
+        line_match = PRAGMA_SETTING_RE.match(line)
+        if line_match:
+            config_lines_detected += 1
+            pv_match = PRAGMA_PV_LINE_RE.search(line)
+            if pv_match:
+                pv_line_detected += 1
+
+    # There shall be at one config line at minimum
+    # There shall be a config line for pv even if it's just "pv:"
+    if (config_lines_detected <= 0 or pv_line_detected <= 0):
```
Comment (klauer) [inline pytmc/bin/pragmalint.py]: Let's break this up into 2 cases, raising the exception with the applicable message from above.

File: pytmc/bin/pragmalint.py [hunk 6]
```diff
@@ -14,16 +14,19 @@
...
 
 PRAGMA_START_RE = re.compile('{attribute')
 PRAGMA_RE = re.compile(
-    r"^{\s*attribute[ \t]+'pytmc'[ \t]*:=[ \t]*'([\s\S]*)'}$",
+    r"^{\s*attribute[ \t]+'pytmc'[ \t]*:=[ \t]*'(?P<setting>[\s\S]*)'}$",
     re.MULTILINE
 )
+PRAGMA_LINE_RE = re.compile(r"(" + _FLEX_TERM_REGEX + ")", re.MULTILINE)
+PRAGMA_SETTING_RE = re.compile(r"^\s*(?P<title>[a-zA-z0-9]+):(?P<setting>.*)$")
+PRAGMA_PV_LINE_RE = re.compile(r"pv\S*:")
```
Comment (klauer) [inline pytmc/bin/pragmalint.py]:
```suggestion
PRAGMA_PV_LINE_RE = re.compile(r"pv\s*:")
```
Lower-case, as it should be whitespace there

File: pytmc/bin/pragmalint.py (hunk 1 above)
Comment (n-wbrown) [reply pytmc/bin/pragmalint.py]:
```suggestion
PRAGMA_SETTING_RE = re.compile(r"^\s*(?P<title>[a-zA-Z0-9]+)\s*:\s*(?P<setting>.*?)\s$")
```

**Comment by n-wbrown** [comment]
Last commit just allows 852 as an acceptable ADS port so the xgmd tests passes. This is a bit of a hack. Travis is failing due to flake8 style issues. I plan to resolve style issues in a future commit.

**Comment by klauer** [comment]
Thanks! 😄

**Pull request merged**